### PR TITLE
Revised enable/disable logic to account for about:preferences#networking checkbox

### PR DESCRIPTION
**Pull request by maxxcrawford**
### Method A: Enable to Disable

1. Add `doh-rollout.enabled` boolean pref, set to `true`
1. On doorhanger notification, select **enable** protections 
1. Go to [about:preferences#general](about:preferences#general) and click "Settings" button in Network Settings section. 
1. Uncheck "Enable DNS over HTTPS" 
1. Go back to [about:debugging](about:debugging#/runtime/this-firefox) and click "Reload" on the add-on. 
1. DoH should be disabled. Confirm by checking about:telemtery or that the `network.trr.mode` pref to confirm it is set to `0`.

### Method B: Disable to Enable

1. Add `doh-rollout.enabled` boolean pref, set to `true`
1. On doorhanger notification, select **disable** protections 
1. Confirm by that the `network.trr.mode` pref to confirm it is set to `5`.
1. Go to [about:preferences#general](about:preferences#general) and click "Settings" button in Network Settings section. 
1. Check "Enable DNS over HTTPS" 
1. Go back to [about:debugging](about:debugging#/runtime/this-firefox) and click "Reload" on the add-on. 
1. DoH should be enabled now. **However, the add-on has no visibility on this, as its been set to a custom setting by the user.**   ~Confirm by checking about:telemtery or that the `network.trr.mode` pref to confirm it is set to `2`.~

### CODE REVIEW

File: src/experiments/doorhanger/api.js [hunk 1]
```diff
@@ -75,9 +76,9 @@ class DoorhangerEventEmitter extends EventEmitter {
       learnMoreURL,
       escAction: "buttoncommand",
       eventCallback: doorhangerEvents,
-      removeOnDismissal: true,
+      removeOnDismissal: false,
     };
-    recentWindow.PopupNotifications.show(browser, "doh-first-time", text, null, primaryAction, secondaryActions, options);
+    let notification = recentWindow.PopupNotifications.show(browser, "doh-first-time", text, null, primaryAction, secondaryActions, options);
```
Comment (jonathanKingston) [inline src/experiments/doorhanger/api.js]: No assignment here needed right?

File: src/experiments/preferences/api.js [hunk 2]
```diff
@@ -35,6 +35,8 @@ ExtensionPreferencesManager.addSetting("dohRollout.state", {
       prefs[TRR_MODE_PREF] = 0;
       break;
     case "manuallyDisabled":
+      prefs[TRR_MODE_PREF] = 0;
```
Comment (jonathanKingston) [inline src/experiments/preferences/api.js]: This is the default now anyway from line 29, maybe we should remove 35 too.

File: src/experiments/doorhanger/api.js (hunk 1 above)
Comment (maxxcrawford) [reply src/experiments/doorhanger/api.js]: This is assigned to close it on the dismissed event _(Line 60 above in the same file)._

`recentWindow.PopupNotifications.remove(notification);`

File: src/experiments/preferences/api.js (hunk 2 above)
Comment (maxxcrawford) [reply src/experiments/preferences/api.js]: I think this has to remain to toggle DoH off during temporary situations (like safe browsing) and back on non-safe browsing networks).

**Comment by maxxcrawford** [comment]
Revising my original test statements to match the new expected logic.

File: src/experiments/doorhanger/api.js (hunk 1 above)
Comment (jonathanKingston) [reply src/experiments/doorhanger/api.js]: Won't the scope cause an issue here? Is it better to declare it before `let doorhangerEvents`?

File: src/experiments/preferences/api.js (hunk 2 above)
Comment (jonathanKingston) [reply src/experiments/preferences/api.js]: Line 29 sets this to 0 anyway is my point, I think it doesn't do anything no?

File: src/background.js [hunk 3]
```diff
@@ -161,12 +173,30 @@ const rollout = {
     await browser.storage.local.set({[name]: value});
   },
 
+  async prefHasUserValueCheck() {
```
Comment (jonathanKingston) [inline src/background.js]: Perhaps add a `trrMode` prefix to this or similar.

File: src/background.js [hunk 4]
```diff
@@ -59,19 +67,22 @@ const stateManager = {
...
-      return false;
-    } else if ( prevMode !== curMode ||  curMode === 5 ||  curMode === 3) {
-      // Add logic specific if user disables DoH in about:config:
-      if ( curMode === 0 ) {
-        await stateManager.setState("manuallyDisabled");
+    // On Mismatch - run never run again (make init check a function)
+
+    if (prevMode !== curMode) {
+      log("Mismatched, curMode: ", curMode);
+      if (curMode === 0) {
```
Comment (jonathanKingston) [inline src/background.js]: Do we care what the value is? Surely 5 and 3 are also important.

Comment (jonathanKingston) [reply src/background.js]: Especially as the else case basically takes care of this within `prefHasUserValueCheck` do we still need anything other than this?

File: src/experiments/preferences/api.js (hunk 2 above)
Comment (maxxcrawford) [reply src/experiments/preferences/api.js]: Ahh, I see it now. Removing both lines!

File: src/background.js (hunk 4 above)
Comment (maxxcrawford) [reply src/background.js]: We have to check for `0` (and apply `stateManager.rememberDisableHeuristics()` bc it's the default right now for network.trr.mode. When the mismatch is detected, we run the `trrModePrefHasUserValueCheck` function, but it doesn't disable heuristics as it appears that there's be no modification to the pref.

File: src/experiments/doorhanger/api.js (hunk 1 above)
Comment (maxxcrawford) [reply src/experiments/doorhanger/api.js]: Revised!

**Pull request merged**